# <div style="text-align: center;">**Assignment 1 – Part 1: Multiprocessing**</div>

In [1]:
# Necessary imports

import time
import random
import multiprocessing
from concurrent.futures import ProcessPoolExecutor

### **Square Program**

In [2]:
def square(n):
    return n * n

# Create a list of 10⁶ random integers (range 1 to 100)
numbers6 = [random.randint(1, 100) for _ in range(10**6)]

## **Test 1: 10⁶ Integers List**

### **Sequential Version**

In [12]:
start_seq = time.time()
results_seq = [square(n) for n in numbers6]
end_seq = time.time()
seq_time = end_seq - start_seq

print(f"First 10 list elements: {numbers6[:10]}")
print(f"\nFirst 10 result elements: {results_seq[:10]}")
print(f"\nSequential execution time for 10⁶ numbers: {seq_time:.4f} seconds")

First 10 list elements: [58, 26, 85, 22, 15, 66, 58, 37, 26, 66]

First 10 result elements: [3364, 676, 7225, 484, 225, 4356, 3364, 1369, 676, 4356]

Sequential execution time for 10⁶ numbers: 0.0549 seconds


### **Multiprocessing Version (Process per Number)**

In [4]:
def worker_square(n, output, index):
    output[index] = square(n)
    
numbers = [random.randint(1, 100) for _ in range(10**6)]

manager = multiprocessing.Manager()
output = manager.list([0] * len(numbers6))
processes = []

start_mp1 = time.time()
for idx, num in enumerate(numbers):
    p = multiprocessing.Process(target=worker_square, args=(num, output, idx))
    processes.append(p)
    p.start()
    
for p in processes:
    p.join()
end_mp1 = time.time()
mp1_time = end_mp1 - start_mp1

print(f"First 10 list elements: {numbers6[:10]}")
print(f"\nFirst 10 result elements: {list(output)[:10]}")
print(f"\nProcess per number multiprocessing execution time for 10⁶ numbers: {mp1_time:.4f} seconds")

Exception ignored in: <function _releaseLock at 0x7fa077ec6e80>
Traceback (most recent call last):
  File "/home/student/anaconda3/envs/parallel/lib/python3.12/logging/__init__.py", line 243, in _releaseLock
    def _releaseLock():
    
KeyboardInterrupt: 


### **Multiprocessing Version (Pooling with map())**

In [9]:
start_mp2 = time.time()
with multiprocessing.Pool() as pool:
    results_map = pool.map(square, numbers6)
end_mp2 = time.time()
mp2_time = end_mp2 - start_mp2

print(f"First 10 list elements: {numbers6[:10]}")
print(f"\nFirst 10 result elements: {list(results_map)[:10]}")
print(f"\nPooling with map() multiprocessing execution time for 10⁶ numbers: {mp2_time:.4f} seconds")

First 10 list elements: [58, 26, 85, 22, 15, 66, 58, 37, 26, 66]

First 10 result elements: [3364, 676, 7225, 484, 225, 4356, 3364, 1369, 676, 4356]

Pooling with map() multiprocessing execution time for 10⁶ numbers: 0.1427 seconds


### **Multiprocessing Version (Pooling with apply())**

In [10]:
start_mp3 = time.time()
results_apply = []
with multiprocessing.Pool() as pool:
    for n in numbers6:
        result = pool.apply(square, (n,))
        results_apply.append(result)
end_mp3 = time.time()
mp3_time = end_mp3 - start_mp3

print(f"First 10 list elements: {numbers6[:10]}")
print(f"\nFirst 10 result elements: {results_map[:10]}")
print(f"\nPooling with apply() multiprocessing execution time for 10⁶ numbers: {mp3_time:.4f} seconds")

First 10 list elements: [58, 26, 85, 22, 15, 66, 58, 37, 26, 66]

First 10 result elements: [3364, 676, 7225, 484, 225, 4356, 3364, 1369, 676, 4356]

Pooling with apply() multiprocessing execution time for 10⁶ numbers: 164.6897 seconds


### **Multiprocessing Version (concurrent futures)**

In [11]:
start_mp4 = time.time()
with ProcessPoolExecutor() as executor:
    results_executor = list(executor.map(square, numbers6))
end_mp4 = time.time()
mp4_time = end_mp4 - start_mp4

print(f"First 10 list elements: {numbers6[:10]}")
print(f"\nFirst 10 result elements: {results_executor[:10]}")
print(f"\nPooling with concurrent futures multiprocessing execution time for 10⁶ numbers: {mp4_time:.4f} seconds")

First 10 list elements: [58, 26, 85, 22, 15, 66, 58, 37, 26, 66]

First 10 result elements: [3364, 676, 7225, 484, 225, 4356, 3364, 1369, 676, 4356]

Pooling with concurrent futures multiprocessing execution time for 10⁶ numbers: 107.7498 seconds


<style>
    .styled-title {
        font-weight: bold;
        font-size: 12px;
        color: #ffffff;
        background: linear-gradient(135deg, #007BFF, #6610f2);
        padding: 5px 10px;
        border-radius: 8px;
        display: inline-block;
    }
</style>

### **Conclusions for Squaring 10⁶ Integers**

<span class="styled-title">Sequential Execution:</span>

  The sequential approach is extremely fast (0.05 sec) for squaring 10⁶ numbers because the computation itself is trivial.

<span class="styled-title">Process per Number:</span>

  Creating one process per number is not feasible due to enormous memory overhead, resulting in a memory allocation error.

<span class="styled-title">Multiprocessing Pool:</span>

  Using **Pool.map()** is the most efficient among the parallel methods tested (0.14 sec), though it’s still slower than the sequential version because of process management overhead.

  Using **Pool.apply()** is highly inefficient (164.69 sec) due to the blocking nature of each call.

<span class="styled-title">Concurrent Futures:</span>

  This method also suffers from high overhead (107.75 sec) compared to **Pool.map()**.

## **Test 2: 10⁷ Integers List**

In [3]:
# Create a list of 10⁷ random integers (range 1 to 100)
numbers7 = [random.randint(1, 100) for _ in range(10**7)]

### **Sequential Version**

In [17]:
start_seq = time.time()
results_seq = [square(n) for n in numbers7]
end_seq = time.time()
seq_time = end_seq - start_seq

print(f"First 10 list elements: {numbers7[:10]}")
print(f"\nFirst 10 result elements: {results_seq[:10]}")
print(f"\nSequential execution time for 10⁷ numbers: {seq_time:.4f} seconds")

First 10 list elements: [18, 8, 60, 11, 78, 1, 35, 11, 72, 17]

First 10 result elements: [324, 64, 3600, 121, 6084, 1, 1225, 121, 5184, 289]

Sequential execution time for 10⁷ numbers: 0.5350 seconds


### **Multiprocessing Version (Process per Number)**

In [ ]:
def worker_square(n, output, index):
    output[index] = square(n)

manager = multiprocessing.Manager()
output = manager.list([0] * len(numbers7))
processes = []

start_mp1 = time.time()
for idx, num in enumerate(numbers7):
    p = multiprocessing.Process(target=worker_square, args=(num, output, idx))
    processes.append(p)
    p.start()
    
for p in processes:
    p.join()
end_mp1 = time.time()
mp1_time = end_mp1 - start_mp1

print(f"First 10 list elements: {numbers7[:10]}")
print(f"\nFirst 10 result elements: {list(output)[:10]}")
print(f"\nProcess per number multiprocessing execution time for 10⁷ numbers: {mp1_time:.4f} seconds")

### **Multiprocessing Synchronous Version (Pooling with map())**

In [18]:
start_mp2 = time.time()
with multiprocessing.Pool() as pool:
    results_map = pool.map(square, numbers7)
end_mp2 = time.time()
mp2_time = end_mp2 - start_mp2

print(f"First 10 list elements: {numbers7[:10]}")
print(f"\nFirst 10 result elements: {list(results_map)[:10]}")
print(f"\nPooling with map() multiprocessing execution time for 10⁷ numbers: {mp2_time:.4f} seconds")

First 10 list elements: [18, 8, 60, 11, 78, 1, 35, 11, 72, 17]

First 10 result elements: [324, 64, 3600, 121, 6084, 1, 1225, 121, 5184, 289]

Pooling with map() multiprocessing execution time for 10⁷ numbers: 1.1927 seconds


### **Multiprocessing Asynchronous Version (Pooling with map())**

In [19]:
start_mp2_async = time.time()
with multiprocessing.Pool() as pool:
    async_result = pool.map_async(square, numbers7)
    results_map_async = async_result.get()
end_mp2_async = time.time()
mp2_async_time = end_mp2_async - start_mp2_async

print(f"First 10 list elements: {numbers7[:10]}")
print(f"\nFirst 10 result elements: {results_map_async[:10]}")
print(f"\nPooling with map_async() multiprocessing execution time for 10⁷ numbers: {mp2_async_time:.4f} seconds")

First 10 list elements: [18, 8, 60, 11, 78, 1, 35, 11, 72, 17]

First 10 result elements: [324, 64, 3600, 121, 6084, 1, 1225, 121, 5184, 289]

Pooling with map_async() multiprocessing execution time for 10⁷ numbers: 1.0862 seconds


### **Multiprocessing Synchronous Version (Pooling with apply())**

In [20]:
start_mp3 = time.time()
results_apply = []
with multiprocessing.Pool() as pool:
    for n in numbers7:
        result = pool.apply(square, (n,))
        results_apply.append(result)
end_mp3 = time.time()
mp3_time = end_mp3 - start_mp3

print(f"First 10 list elements: {numbers7[:10]}")
print(f"\nFirst 10 result elements: {results_map[:10]}")
print(f"\nPooling with apply() multiprocessing execution time for 10⁷ numbers: {mp3_time:.4f} seconds")

First 10 list elements: [18, 8, 60, 11, 78, 1, 35, 11, 72, 17]

First 10 result elements: [324, 64, 3600, 121, 6084, 1, 1225, 121, 5184, 289]

Pooling with apply() multiprocessing execution time for 10⁷ numbers: 1684.7950 seconds


### **Multiprocessing Asynchronous Version (Pooling with apply())**

In [21]:
start_mp3_async = time.time()
with multiprocessing.Pool() as pool:
    async_results = [pool.apply_async(square, (n,)) for n in numbers7]
    results_apply_async = [r.get() for r in async_results]
end_mp3_async = time.time()
mp3_async_time = end_mp3_async - start_mp3_async

print(f"First 10 list elements: {numbers7[:10]}")
print(f"\nFirst 10 result elements: {results_apply_async[:10]}")
print(f"\nPooling with apply_async() multiprocessing execution time for 10⁷ numbers: {mp3_async_time:.4f} seconds")

First 10 list elements: [18, 8, 60, 11, 78, 1, 35, 11, 72, 17]

First 10 result elements: [324, 64, 3600, 121, 6084, 1, 1225, 121, 5184, 289]

Pooling with apply_async() multiprocessing execution time for 10⁷ numbers: 533.6804 seconds


### **Multiprocessing Version (concurrent futures)**

In [4]:
start_mp4 = time.time()
with ProcessPoolExecutor() as executor:
    results_executor = list(executor.map(square, numbers7))
end_mp4 = time.time()
mp4_time = end_mp4 - start_mp4

print(f"First 10 list elements: {numbers7[:10]}")
print(f"\nFirst 10 result elements: {results_executor[:10]}")
print(f"\nPooling with concurrent futures multiprocessing execution time for 10⁷ numbers: {mp4_time:.4f} seconds")

First 10 list elements: [18, 96, 11, 97, 88, 3, 28, 97, 35, 6]

First 10 result elements: [324, 9216, 121, 9409, 7744, 9, 784, 9409, 1225, 36]

Pooling with concurrent futures multiprocessing execution time for 10⁷ numbers: 1099.2878 seconds


<style>
    .styled-title {
        font-weight: bold;
        font-size: 12px;
        color: #ffffff;
        background: linear-gradient(135deg, #007BFF, #6610f2);
        padding: 5px 10px;
        border-radius: 8px;
        display: inline-block;
    }

    .final-conclusion {
        font-weight: bold;
        font-size: 14px;
        color: #ffffff;
        background-color: #007BFF;
        padding: 10px;
        border-radius: 8px;
        display: inline-block;
        margin-top: 10px;
    }
</style>

### **Conclusions for Squaring 10⁷ Integers**

<span class="styled-title">Sequential Execution:</span>

  The sequential approach is also extremely fast (0.54 sec) for squaring 10⁷ numbers because the computation itself is trivial and avoids parallel overhead.

<span class="styled-title">Process per Number:</span>

  Creating one process per number remains unfeasible due to extreme memory overhead, resulting in an **OSError (memory allocation failure)**. This approach is highly inefficient.

<span class="styled-title">Multiprocessing Pool:</span>

  Using **Pool.map()** is again the most efficient parallel approach (1.19 sec), though it is still slower than sequential execution due to process management overhead.

  Using **Pool.map_async()** slightly improves performance compared to the synchronous mapping (1.09 sec), benefiting from asynchronous execution.

  Using **Pool.apply()** is highly inefficient (1684.80 sec) because it blocks execution for each number, making it significantly slower than all other methods.

  Using **Pool.apply_async()** is an improvement over **apply()** (533.68 sec) but still much slower than **map()** due to individual function calls.

<span class="styled-title">Concurrent Futures:</span>

  This method also suffers from high overhead (1099.29 sec) compared to **Pool.map()**, making it inefficient for such a simple computation.

<span class="final-conclusion">**Final Conclusion:**</span>

For squaring large numbers, **sequential execution remains the fastest** due to minimal overhead. **Pool.map()** is the best parallel approach, but multiprocessing generally adds unnecessary complexity for simple arithmetic operations.